### ISSUE TITLE
How do I max out CPU cores with AutoML?

### ISSUE THREAD

**Issue by jaxidian** [dotnet/machinelearning#4092]
I'm using AutoML to find and train the "best model possible" for my set of data. This is awesome and works really well - thanks all!

My question/problem comes to trying to max out the CPU resources when doing this resource-intensive process. I have a workstation with 12 cores (24 threads) yet when I have AutoML hammering at my system for multiple minutes, I only see ~20-25% CPU usage with most of my cores sitting idle. Given that it is training many different models at once, shouldn't it be able to make use of every single core available, complete with SMT resources as well? Is this configurable somewhere?

**Comment by CESARDELATORRE** [comment]
Adding @justinormont @vinodshanbhag @Dmitry-A from the AutoML team. 

Thanks a lot for the feedback! 👍

**Comment by justinormont** [comment]
The CPU utilization will depend on the trainer running.

Few optimizations we should make:
* Parallel sweeping - https://github.com/dotnet/machinelearning-automl/issues/393
* Set NumberOfThreads in each trainer - https://github.com/dotnet/machinelearning-automl/issues/394
* Cache featurized dataset before trainer and/or cache trained transforms (for re-use between pipelines) - https://github.com/dotnet/machinelearning-automl/issues/396

Apologies for the links to a non-public repo. We still have to transfer the issues to the current repo.

**Comment by jaxidian** [comment]
Hi @justinormont,

All of those links are 404s for me, I assume due to lack of a permission. Without being able to see any of those details and making some wild assumptions, I suspect parallel sweeping might be the lowest-hanging fruit since I'm primarily wanting to increase the number of models evaluated within N seconds (I'm getting ~1 model per 5 seconds on what I think is a single core, of course with some significant variance in there). Searching online for info on "Parallel Sweeping" is turning up almost nothing for me.

**Comment by jaxidian** [comment]
In case it's useful to see, this is the base code I'm using now that I'd like to enhance:

            var experimentSettings = new RegressionExperimentSettings
            {
                MaxExperimentTimeInSeconds = (uint)Math.Max(trainingDurationInSeconds, 10),
                OptimizingMetric = RegressionMetric.RootMeanSquaredError
            };

            var mlContext = new MLContext();
            var rawTrainingData = await Repository.GetTrainingData(forecastGroupId);
            var transformedTrainingData = rawTrainingData.Select(x => Mapper.Map<HistoryEntity, HistoryModel>(x));
            var trainingDataView = mlContext.Data.LoadFromEnumerable(transformedTrainingData);
            var experiment = mlContext.Auto().CreateRegressionExperiment(experimentSettings);
            var experimentResult = experiment.Execute(trainingDataView,
                labelColumnName: nameof(HistoryEntity.Quantity));

            var filteredResulingMetrics = experimentResult.RunDetails.Where(x => x.ValidationMetrics.RSquared > 0.1); // For some reason, we sometimes get a run with zeroes for all metrics

**Comment by SnakyBeaky** [comment]
I'm guessing a quick/dirty solution for this would be to fire several regression experiments at once, each on a different thread and excluding trainers from one of the experiments in the rest.

Example diving all regression trainers into two sets running in parallel.

```csharp
RegressionExperimentSettings experimentSettingsOne = new RegressionExperimentSettings();
experimentSettings.Trainers.Remove(RegressionTrainer.LightGbm);
experimentSettings.Trainers.Remove(RegressionTrainer.Ols);
experimentSettings.Trainers.Remove(RegressionTrainer.OnlineGradientDescent);
experimentSettings.Trainers.Remove(RegressionTrainer.StochasticDualCoordinateAscent);

RegressionExperiment experimentOne = Context.Auto().CreateRegressionExperiment(experimentSettingsOne);

RegressionExperimentSettings experimentSettingsTwo = new RegressionExperimentSettings();
experimentSettings.Trainers.Remove(RegressionTrainer.FastForest);
experimentSettings.Trainers.Remove(RegressionTrainer.FastTree);
experimentSettings.Trainers.Remove(RegressionTrainer.FastTreeTweedie);
experimentSettings.Trainers.Remove(RegressionTrainer.LbfgsPoissonRegression);

RegressionExperiment experimentTwo = Context.Auto().CreateRegressionExperiment(experimentSettingsTwo);

Task<ExperimentResult<RegressionMetrics>> firstExperiment = Task.Run(() => experimentOne.Execute(trainingData));

Task<ExperimentResult<RegressionMetrics>> secondExperiment = Task.Run(() => experimentTwo.Execute(trainingData));

await Task.WhenAll(firstExperiment, secondExperiment);

ExperimentResult<RegressionMetrics> firstResults = await firstExperiment;
ExperimentResult<RegressionMetrics> secondResults = await secondExperiment;
```

**Comment by luisquintanilla** [comment]
Closing this since it looks like it was addressed  in dotnet/machinelearning-modelbuilder#645 and relates to the previous implementation of AutoML.
